nixos/modules/tasks/filesystems/overlayfs.nix

```diff
@@ -0,0 +1,147 @@
+{ config, lib, pkgs, utils, ... }:
+
+let
+  # The scripted initrd contains some magic to add the prefix to the
+  # paths just in time, so we don't add it here.
+  sysrootPrefix = fs:
+    if config.boot.initrd.systemd.enable && (utils.fsNeededForBoot fs) then
+      "/sysroot"
+    else
+      "";
+
+  # Returns a service that creates the required directories before the mount is
+  # created.
+  preMountService = _name: fs:
```

Comment:
My first question was how do we make sure this service is executed before the mount? It's not clear to me.

However then I just also had a shower-thought. Mount units are `systemd.exec` units; so instead of using a service that runs before the mount we could just add

```
mounts.${utils.escapeSystemdPath fs.mountPoint}.mountConfig.StateDirectory = [ "upperdir" "workdir" ];
```

and then have the upperdir and workdir settings be `/var/lib/${mountName}/{upperdir,workdir}`

Then we don't need a separate service for this

Reply:
The service will execute first because of the `x-systemd.after=` and `x-systemd.requires=` options on the file system.

Using a `StateDirectory` is *interesting*, but only really works if that's actually where the user wants that state stored...

Reply:
`RuntimeDirectory=,CacheDirectory=` and friends are also available

Reply:
The ownership-changing behaviour implied by these options seems _very_ undesirable here.

Additionally, I don't think this allows a reasonable interface for specifying the filesystem that the upper/work directories are supposed to live on?